### PR TITLE
PFDR-98 - Generalize and use policy mocks throughout controllers

### PR DESCRIPTION
**Pull request by botimer**
This is the beginning of a PR for PFDR-98. It introduces a pattern and uses it in one place. There is plenty of work to do to apply it elsewhere. Flagging this as draft for now.

### CODE REVIEW

**Comment by aelkiss** [comment]
In general I'm pretty happy with this, and I think it meets the use case for PFDR-98 pretty well especially in the specs.

What if we spiced things up in the controllers with a soupçon of metaprogramming:

```ruby

class ResourceController
  attr_writer :collection_policy
  attr_writer :resource_policy

  def self.resource_policy(policy)
    define_method(:resource_policy) { @resource_policy ||= policy }
  end

  def self.collection_policy(policy)
    define_method(:collection_policy) { @collection_policy ||= policy }
  end

  resource_policy RejectAll
  collection_policy RejectAll
end

class PackageController < ResourceController
  resource_policy PackagePolicy
  collection_policy PackagesPolicy

  ...
end
```

**Comment by botimer** [comment]
I'm favorable toward the occasional class method for declarative purposes, particularly where you can find the thing you're calling and it has very simple behavior. I think I would probably do this with a class variable rather (coalescing to, but not assigning, RejectAll) than a dynamic method:

```
class ResourceController
  def self.set_resource_policy(policy)
    @@resource_policy = policy
  end

  def resource_policy
    @@resource_policy || RejectAll
  end
end
```

File: spec/controllers/v1/packages_controller_spec.rb [hunk 1]
```diff
@@ -3,32 +3,93 @@
...
+    it_behaves_like "an index endpoint", "PackagesPolicy"
+
     describe "GET #show" do
-      it_behaves_like "a show endpoint" do
-        let(:policy) { PackagePolicy }
-        let(:key) { :bag_id }
-        let(:factory) do
-          proc {|user| user ? Fabricate(:package, user: user) : Fabricate(:package) }
+      context "when the resource belongs to the user" do
+        include_context "as underprivileged user"
```
Comment (aelkiss) [inline spec/controllers/v1/packages_controller_spec.rb]: with someone logged in

File: spec/controllers/v1/packages_controller_spec.rb [hunk 2]
```diff
@@ -3,32 +3,93 @@
...
 
+    it_behaves_like "an index endpoint", "PackagesPolicy"
+
     describe "GET #show" do
-      it_behaves_like "a show endpoint" do
-        let(:policy) { PackagePolicy }
-        let(:key) { :bag_id }
-        let(:factory) do
-          proc {|user| user ? Fabricate(:package, user: user) : Fabricate(:package) }
+      context "when the resource belongs to the user" do
```
Comment (aelkiss) [inline spec/controllers/v1/packages_controller_spec.rb]: when the policy allows the user to see the resource

File: spec/controllers/v1/packages_controller_spec.rb [hunk 3]
```diff
@@ -3,32 +3,93 @@
...
+        include_context "as underprivileged user"
+
+        it "raises an ActiveRecord::RecordNotFound" do
+          expect do
+            get :show, params: { bag_id: '(missing)' }
+          end.to raise_exception ActiveRecord::RecordNotFound
+        end
+      end
+
+      context "as an unauthorized user" do
```
Comment (aelkiss) [inline spec/controllers/v1/packages_controller_spec.rb]: parallel change to context about the policy above

File: spec/controllers/v1/packages_controller_spec.rb [hunk 4]
```diff
@@ -3,32 +3,93 @@
...
+        let(:package) { Fabricate(:package, user: user) }
+
+        it "can fetch a package by external id" do
+          resource_policy 'PackagePolicy', show?: true
+          get :show, params: { bag_id: package.external_id }
+
+          expect(assigns(:package)).to eql(package)
+        end
+
+        it "rejects unauthorized users" do
```
Comment (aelkiss) [inline spec/controllers/v1/packages_controller_spec.rb]: same change as above needed

File: spec/support/helpers/policy_double.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+#
+# The simplest construct is to supply only a hash of predicates and whether
+# they should return true or false. The double returned will expect that
+# authorize! is called and will raise unless the predicate is truthy.
+#
+# Example:
+#
+# let(:policy) { policy_double(show?: true) }
+#
+# It does not expect that the specific predicate is authorized, but the
```
Comment (aelkiss) [inline spec/support/helpers/policy_double.rb]: Ensure this comment reflects the actual expectations that are set up

File: spec/support/helpers/policy_double.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+# the authorize! expectation is not set automatically and the block is called
+# each time a new policy instance is created. The first parameter is the double
+# itself and the rest are what is passed to the policy constructor. You can
+# set more detailed responses and expectations on the policy instance here.
+#
+# Example:
+#
+# controller.resource_policy = policy_double do |policy|
+#   allow(policy).to receive(:show?).and_return(true)
+#   expect(policy).to receive(:authorize!).with(:show?)
```
Comment (aelkiss) [inline spec/support/helpers/policy_double.rb]: Would be nice to have an example of a block with expectations about the additional parameters (user, resource, scope, etc)

**Comment by botimer** [comment]
@aelkiss All comments are addressed. We could add a more exotic example for the block form of `policy_double`, but I expanded it to make a small expectation with user/resource. I think that's probably good enough until we extract this to Checkpoint and have concrete examples.

Ready for final approval/merge.

**Pull request merged**